feat(routing/http): IPIP-518: HTTP(S) URLs in Routing V1 API

lidel opened the pull request:
This PR implements [IPIP-518](https://github.com/ipfs/specs/pull/518) which extends `/routing/v1` with support for URIs alongside multiaddrs.

## Changes

### New `Address` type (`types/address.go`)

- accepts both multiaddrs (starting with `/`) and URIs, following IPIP-518 parsing rules
- schema-agnostic: any valid URI scheme is accepted for future extensibility
- protocol matching with `HasProtocol` handles http/https semantics (e.g. `https://` matches both "http" and "https" queries)
- `ToMultiaddr()` converts HTTP(S) URLs to multiaddrs for backward compatibility with existing software that expects `/dns/host/tcp/port/https` format, including `/http-path/` when the URL has a non-empty path
  - paths are not used  (consumed) by Kubo atm, but this will make WebSeed integration easier in the future, i hope
- invalid addresses are skipped per spec, not treated as errors

### New `GenericRecord` type (`types/record_generic.go`)

- schema-agnostic record type using `Addresses` (which accepts both URLs and multiaddrs) instead of `[]Multiaddr`
- preserves unknown fields in `Extra` map for forward compatibility with future protocols
- size-limited to 10 KiB to prevent abuse in NDJSON streams

### Contentrouter integration (`contentrouter/contentrouter.go`)

- `readProviderResponses` converts `GenericRecord` to `peer.AddrInfo` for legacy routing APIs:
  - records with valid PeerID: always converted (supports legacy PeerID + `/https` multiaddr pattern)
  - records with `transport-ipfs-gateway-http` + HTTP(S) URLs: PeerID derived from `did:key:` (ed25519) or generated as a deterministic placeholder
  - other non-PeerID records: skipped
- PeerID interop helpers extracted to `contentrouter/interop.go`

### Filter updates (`filters/filters.go`)

- `ApplyAddrFilter` extended to handle `Address` type alongside `Multiaddr`
- protocol filtering works across both URL schemes and multiaddr protocols

### Existing types unchanged

- `PeerRecord` and `BitswapRecord` continue to use `[]Multiaddr` (multiaddrs only)
- URLs belong in `GenericRecord`, which is the forward-looking record type

## Backward compatibility

- multiaddr-only implementations continue to work unchanged
- HTTP(S) URLs are converted to multiaddrs via `ToMultiaddr()` so existing consumers (Kubo, someguy) receive addresses they can use
- `PeerRecord.Addrs` remains `[]Multiaddr` and rejects URLs at deserialization time

IPIP:

- https://github.com/ipfs/specs/pull/518

## TODO

- [ ] no objections in https://github.com/ipfs/specs/pull/518
- [ ] PR in someguy that uses this PR and tests there pass green

lidel committed:
feat(routing/http): implement IPIP-518 HTTP(S) URLs in Delegated Routing API

Implements schema-agnostic support for HTTP(S) URLs alongside multiaddrs
in the Delegated Routing V1 HTTP API as specified in IPIP-518.

Key changes:
- Add new Address type that accepts both multiaddrs (starting with /) and URIs
- Update PeerRecord and BitswapRecord to use Addresses type instead of []Multiaddr
- Implement defensive programming: unsupported addresses are skipped, not errors
- Add special protocol filtering for http/https/tls matching
- Maintain full backward compatibility with multiaddr-only implementations

The implementation is schema-agnostic, accepting any valid URI scheme for
future extensibility. URLs are preserved as-is without lossy conversion
to multiaddrs, avoiding conversion issues documented in multiaddr#63.

Ref: https://github.com/ipfs/specs/pull/518

lidel committed:
feat(routing/http): add URL-to-multiaddr conversion

Implements backward compatibility layer for HTTP(S) URL to multiaddr
conversion during transition period. Uses /https (not /tls/http) to
match IPNI convention and /dns for dual-stack support.

- Add ToMultiaddr() method and String() for Addresses type
- Replace TODOs in contentrouter with actual conversion
- Add comprehensive tests

Temporary compatibility layer until ecosystem adopts IPIP-518 URLs.

lidel committed:
docs(routing/http): document ToMultiaddr in CHANGELOG

gammazero commented on routing/http/client/client_test.go:
```diff
@@ -146,18 +146,21 @@ func makeCID() cid.Cid {
 }
 
-func addrsToDRAddrs(addrs []multiaddr.Multiaddr) (drmas []types.Multiaddr) {
+func addrsToDRAddrs(addrs []multiaddr.Multiaddr) types.Addresses {
```
```suggestion
func multiaddrsToRoutingAddrs(addrs []multiaddr.Multiaddr) types.Addresses {
```

gammazero commented on routing/http/types/address.go:
```diff
@@ -0,0 +1,319 @@
+func (a *Address) Protocols() []string {
+	if a.url != nil {
+		return []string{a.url.Scheme}
+	} else if a.multiaddr != nil {
```
```suggestion
	}
	if a.multiaddr != nil {
```

gammazero commented on routing/http/types/address.go:
```diff
@@ -0,0 +1,319 @@
+		case "tls":
+			// "tls" matches any multiaddr with /tls
+			for _, p := range protocols {
+				if p == "tls" {
```
Should this also match "https"?

lidel replied:
i think "tls" is bit too vague, remove it for now

gammazero commented on routing/http/types/address.go:
```diff
@@ -0,0 +1,319 @@
+	} else {
+		// For HTTP, use /http
+		maStr = fmt.Sprintf("/%s/%s/tcp/%s/http", addrProto, host, port)
+	}
```
Since scheme can only be "http" or "https" just use scheme, instead of the if-else.
```suggestion
	// For HTTPS, use /https as this is what existing HTTP providers
	// announce on IPNI, so we follow same convention for backward-compatibility
	maStr := fmt.Sprintf("/%s/%s/tcp/%s/%s", addrProto, host, port, scheme)
```

lidel committed:
refactor(routing/http): pivot IPIP-518 to generic schema with PeerID compat

introduce GenericRecord (Schema=generic) as specified in the updated
IPIP-518, keeping PeerRecord (Schema=peer) unchanged with []Multiaddr.

type system:
- add GenericRecord with duck-typed Addresses (multiaddrs + URIs)
- revert PeerRecord.Addrs and BitswapRecord.Addrs to []Multiaddr
- change Address constructors to return value types instead of pointers
- remove special "tls" matching from HasProtocol (redundant with default)
- simplify ToMultiaddr to use URL scheme directly (http or https)

contentrouter (client-side GenericRecord -> peer.AddrInfo conversion):
- convert GenericRecord to peer.AddrInfo for legacy libp2p routing APIs
- if ID is a valid PeerID, convert regardless of Protocols (supports the
  legacy pattern where PeerID + /https multiaddr hints at a Trustless
  IPFS HTTP Gateway even without explicit protocol declaration)
- if ID is not a PeerID but record has transport-ipfs-gateway-http +
  HTTP(S) URLs, derive PeerID from did:key: or generate a placeholder
- HTTPS URLs become /dns/host/tcp/443/https multiaddrs via ToMultiaddr
- extract filterAddrs helper for nil-safe []Multiaddr extraction
- add MaxGenericRecordSize check in JSON deserialization path to match
  the existing NDJSON check

filters (server-side):
- add GenericRecord support to ApplyFiltersToIter
- add applyGenericFilters for protocol and address filtering on
  GenericRecord (operates on Addresses directly, no conversion needed)
- convert PeerRecord []Multiaddr to Addresses for filtering via
  applyAddrFilter, then convert back

server and client:
- serialize/deserialize GenericRecord in JSON and NDJSON responses
- add GenericRecord to FindProviders end-to-end client test

tests:
- add GenericRecord JSON/NDJSON round-trip tests with Extra field
  preservation, unparseable address handling, and size limit enforcement
- add contentrouter tests for PeerID, did:key, placeholder PeerID,
  empty Protocols, and HTTP URL conversion paths
- add filter tests for mixed multiaddr+URL addresses and mixed
  PeerRecord+GenericRecord iterator streams
- add server tests for mixed schema responses, protocol/addr filtering,
  NDJSON streaming, and Extra field serialization
- document that PeerRecord Addrs rejects URLs at unmarshal time

lidel committed:
Merge remote-tracking branch 'origin/main' into ipip-518-urls-in-routing

# Conflicts:
#	CHANGELOG.md
#	routing/http/contentrouter/contentrouter.go
#	routing/http/filters/filters.go

lidel committed:
refactor(routing/http): preserve http-path in ToMultiaddr, extract interop helpers

- ToMultiaddr() now appends /http-path/ when the URL has a non-empty path,
  matching go-multiaddr http-path protocol conventions
- moved peerIDFromDIDKey, peerIDPlaceholderFromArbitraryID, hasHTTPURL
  to contentrouter/interop.go for better organization
- added unit tests for hasHTTPURL covering http, https, non-HTTP,
  multiaddr, mixed, empty, and invalid address cases
- fixed staticcheck lint suppression for SchemaBitswap in ndjson test
